rfcs/20240315-algorithm-support.md

```diff
@@ -0,0 +1,166 @@
+  accumulator type is F32.
+* `BF16_BF16_F32_X3`: An algorithm which uses three BF16_BF16_F32 dot operations
+  per "tile" to achive a higher precision.
+* `BF16_BF16_F32_X6`: An algorithm which uses six BF16_BF16_F32 dot operations
+  per "tile" to achive a higher precision ("similar" to F32_F32_F32).
+* `TF32_TF32_F32`: The inputs are casted to TF32 precision and the accumulator
+  type is F32.
+* `TF32_TF32_F32_X3`: An algorithm which uses 3 TF32_TF32_F32 matmuls to achieve
+  a higher precision ("similar" to F32_F32_F32).
+* `F32_F32_F32`: The inputs are used with / casted to F32 precision and the
+  accumulator type is also F32.
+* `F64_F64_F64`: The inputs are used with / casted to F64 precision and the
+  accumulator type is also F64.
+
+In general, it is not guaranteed that the each algorithm is supported on each
+accelerator type by the consumer of the StableHLO. If a given algorithm is not
+supported, an error should be raised as opposed to falling back to an
+alternative.
+
+*Note: In the code, the enum defs would be prefixed with
+STABLEHLO_DOT_ALGORITHM_.*
+
+#### Inputs
+
+*Add this row to the table:*
+
+| Label | Name        | Type                                                                                                                                                                                                                                                    | Constraints |
+| ----- | ----------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------- |
+| (I8)  | `algorithm` | an enum of `UNSET`, `ANY_F8_ANY_F8_F32`, `ANY_F8_ANY_F8_F32_FAST_ACCUM`, `F16_F16_F16`, `F16_F16_F32`, `BF16_BF16_BF16`, `BF16_BF16_F32`, `BF16_BF16_F32_X3`, `BF16_BF16_F32_X6`, `TF32_TF32_F32`, `TF32_TF32_F32_X3`, `F32_F32_F32`, and `F64_F64_F64` |             |
+
+#### Examples
+
+*Change the example:*
+
+```mlir
+// %lhs: [
+//        [[1, 2],
+//         [3, 4]],
+//        [[5, 6],
+//         [7, 8]]
+//       ]
+// %rhs: [
+//        [[1, 0],
+//         [0, 1]],
+//        [[1, 0],
+//         [0, 1]]
+//       ]
+%result = "stablehlo.dot_general"(%lhs, %rhs) {
+  dot_dimension_numbers = #stablehlo.dot<
+    lhs_batching_dimensions = [0],
+    rhs_batching_dimensions = [0],
+    lhs_contracting_dimensions = [2],
+    rhs_contracting_dimensions = [1]
+  >,
+  precision_config = [#stablehlo<precision DEFAULT>, #stablehlo<precision DEFAULT>],
+  algorithm = #stablehlo<dot_algorithm UNSET>
+} : (tensor<2x2x2xi64>, tensor<2x2x2xi64>) -> tensor<2x2x2xi64>
+// %result: [
+//           [[1, 2],
+//            [3, 4]],
+//           [[5, 6],
+//            [7, 8]]
+//          ]
+```
+
+## Alternatives considered
+
+### Adding new Precision values instead of the algorithm property
```

Comment:
I'm debating on this alternative.

On one hand it's super nice to not introduce a separate attribute, on the other, I like that we don't need an array when specifying an algorithm.

I think I still prefer this alternative, if we go this route we also don't need a sentinel enum value for `<UNSET>` since that can be represented by empty array. We can handle per-op discrepancy in op-verification pretty easily.

Reply:
I think we can do what you recommended for now.
(In XLA, we used a separate algorithm flag, but I think we can handle this difference at the conversion from StableHLO to HLO.)
It should work for now and if anyone needs something more complex we can iterate further.

Reply:
The resolution is to reuse the precision property instead of adding a new algorithm property.
But we still have to add new enum values for precision.

Reply:
@reedwm FYI. I think you (also) advocated for the new algorithm field, because it's nicer than reusing the operand precisions, but Kevin said that it's easier to get this RFC accepted if we just reuse it.
I guess it doesn't matter that much anyway. Also it may later better support algorithms where we set different precision for the 2 operands of the matmul.